File: src/compiler.jl
# MinRust Compiler - Integration Module
# 모든 컴파일 단계를 통합

module MinRustCompiler

using ..MinRustTokenizer
using ..MinRustParser
using ..MinRustTypeChecker
using ..MinRustCodeGen

export compile, CompilationResult

# ============================================================
# Compilation Result
# ============================================================

struct CompilationResult
    success::Bool
    c_code::Union{String, Nothing}
    errors::Vector{String}
    token_count::Int
    ast_nodes::Int
    warnings::Vector{String}
end

function Base.show(io::IO, result::CompilationResult)
    if result.success
        println(io, "✅ Compilation Success")
        println(io, "  Tokens: $(result.token_count)")
        println(io, "  AST Nodes: $(result.ast_nodes)")
        if !isempty(result.warnings)
            println(io, "  Warnings: $(length(result.warnings))")
        end
    else
        println(io, "❌ Compilation Failed")
        println(io, "  Errors: $(length(result.errors))")
        for error in result.errors
            println(io, "    - $error")
        end
    end
end

# ============================================================
# Main Compilation Pipeline
# ============================================================

function compile(input::String)::CompilationResult
    errors = String[]
    warnings = String[]

    # Step 1: Tokenization
    try
        tokens = tokenize(input)
        if isempty(tokens)
            push!(errors, "Tokenizer produced no tokens")
            return CompilationResult(false, nothing, errors, 0, 0, warnings)
        end

        # Step 2: Parsing
        try
            parser = MinRustParser.Parser(tokens)
            program = MinRustParser.parse(tokens)

            if program === nothing || isempty(program.items)
                push!(errors, "Parser produced empty AST")
                return CompilationResult(false, nothing, errors, length(tokens), 0, warnings)
            end

            ast_node_count = count_ast_nodes(program)

            # Step 3: Type Checking
            try
                type_check_result = MinRustTypeChecker.check_types(program)

                if !type_check_result.success
                    for error in type_check_result.errors
                        push!(errors, error.message)
                    end
                    return CompilationResult(false, nothing, errors, length(tokens), ast_node_count, warnings)
                end

                # Step 4: Code Generation
                try
                    c_code = MinRustCodeGen.generate_code(program)

                    return CompilationResult(
                        true,
                        c_code,
                        errors,
                        length(tokens),
                        ast_node_count,
                        warnings
                    )

                catch e
                    push!(errors, "Code generation error: $(e)")
                    return CompilationResult(false, nothing, errors, length(tokens), ast_node_count, warnings)
                end

            catch e
                push!(errors, "Type checking error: $(e)")
                return CompilationResult(false, nothing, errors, length(tokens), ast_node_count, warnings)
            end

        catch e
            push!(errors, "Parsing error: $(e)")
            return CompilationResult(false, nothing, errors, length(tokens), 0, warnings)
        end

    catch e
        push!(errors, "Tokenization error: $(e)")
        return CompilationResult(false, nothing, errors, 0, 0, warnings)
    end
end

# ============================================================
# Helper Functions
# ============================================================

function count_ast_nodes(program::MinRustAST.Program)::Int
    count = 1  # Program itself

    for item in program.items
        count += count_node(item)
    end

    return count
end

function count_node(node::Any)::Int
    count = 1

    if isa(node, MinRustAST.FnDecl)
        count += count_node(node.body)
        for param in node.params
            count += 1
        end
    elseif isa(node, MinRustAST.Block)
        for stmt in node.statements
            count += count_node(stmt)
        end
    elseif isa(node, MinRustAST.IfStmt)
        count += count_node(node.condition)
        count += count_node(node.then_block)
        if node.else_block !== nothing
            count += count_node(node.else_block)
        end
    elseif isa(node, MinRustAST.ForStmt)
        count += count_node(node.iter_expr)
        count += count_node(node.body)
    elseif isa(node, MinRustAST.WhileStmt)
        count += count_node(node.condition)
        count += count_node(node.body)
    elseif isa(node, MinRustAST.BinaryOp)
        count += count_node(node.left)
        count += count_node(node.right)
    elseif isa(node, MinRustAST.UnaryOp)
        count += count_node(node.operand)
    elseif isa(node, MinRustAST.Call)
        for arg in node.args
            count += count_node(arg)
        end
    end

    return count
end

# ============================================================
# Compilation with Custom Options
# ============================================================

struct CompilerOptions
    optimize::Bool
    debug::Bool
    output_ir::Bool
end

function compile_with_options(input::String, options::CompilerOptions)::CompilationResult
    result = compile(input)

    if options.output_ir
        # 미래: IR 출력 옵션
    end

    if options.optimize
        # 미래: 최적화 적용 (Phase 6)
    end

    return result
end

# ============================================================
# Compilation Pipeline Statistics
# ============================================================

function print_compilation_stats(result::CompilationResult)
    println("═══════════════════════════════════════")
    println("MinRust Compilation Statistics")
    println("═══════════════════════════════════════")
    println("Status: $(result.success ? "✅ Success" : "❌ Failed")")
    println("Tokens: $(result.token_count)")
    println("AST Nodes: $(result.ast_nodes)")

    if !isempty(result.errors)
        println("\nErrors: $(length(result.errors))")
        for (i, error) in enumerate(result.errors)
            println("  [$i] $error")
        end
    end

    if !isempty(result.warnings)
        println("\nWarnings: $(length(result.warnings))")
        for (i, warning) in enumerate(result.warnings)
            println("  [$i] $warning")
        end
    end

    if result.success && result.c_code !== nothing
        c_lines = length(split(result.c_code, '\n'))
        println("\nGenerated C Code:")
        println("  Lines: $c_lines")
    end

    println("═══════════════════════════════════════")
end

end  # module MinRustCompiler

File: src_rust/src/compiler.rs
// MinRust Compiler - Phase 5 Rust Implementation
// Main compilation pipeline orchestrator

use crate::tokenizer::Tokenizer;
use crate::parser::Parser;
use crate::type_checker::TypeChecker;
use crate::codegen::CodeGenerator;

/// Compilation result
#[derive(Debug, Clone)]
pub struct CompilationResult {
    pub success: bool,
    pub c_code: Option<String>,
    pub errors: Vec<String>,
    pub warnings: Vec<String>,
    pub token_count: usize,
    pub ast_node_count: usize,
}

impl CompilationResult {
    /// Print compilation statistics
    pub fn print_stats(&self) {
        println!("Compilation Result:");
        println!("  Success: {}", self.success);
        println!("  Tokens: {}", self.token_count);
        println!("  AST Nodes: {}", self.ast_node_count);
        println!("  Errors: {}", self.errors.len());
        println!("  Warnings: {}", self.warnings.len());

        if !self.errors.is_empty() {
            println!("\nErrors:");
            for error in &self.errors {
                println!("  - {}", error);
            }
        }

        if !self.warnings.is_empty() {
            println!("\nWarnings:");
            for warning in &self.warnings {
                println!("  - {}", warning);
            }
        }
    }
}

/// Main compiler
pub struct Compiler;

impl Compiler {
    /// Compile Rust source to C code
    pub fn compile(source: &str) -> CompilationResult {
        let mut result = CompilationResult {
            success: false,
            c_code: None,
            errors: Vec::new(),
            warnings: Vec::new(),
            token_count: 0,
            ast_node_count: 0,
        };

        // Stage 1: Tokenization
        let mut tokenizer = Tokenizer::new(source);
        let tokens = tokenizer.tokenize();
        result.token_count = tokens.len();

        // Stage 2: Parsing
        let mut parser = Parser::new(tokens);
        match parser.parse() {
            Ok(program) => {
                result.ast_node_count = Self::count_ast_nodes(&program);

                // Stage 3: Type Checking
                let mut type_checker = TypeChecker::new();
                match type_checker.check(&program) {
                    Ok(_) => {
                        // Stage 4: Code Generation
                        let mut codegen = CodeGenerator::new();
                        let c_code = codegen.generate(&program);
                        result.c_code = Some(c_code);
                        result.success = true;
                    }
                    Err(errors) => {
                        for error in errors {
                            result.errors.push(error.message);
                        }
                    }
                }
            }
            Err(errors) => {
                for error in errors {
                    result.errors.push(error.message);
                }
            }
        }

        result
    }

    /// Count AST nodes (placeholder)
    fn count_ast_nodes(program: &crate::ast::Program) -> usize {
        program.items.len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_compile_empty() {
        let result = Compiler::compile("");
        assert!(result.success || result.c_code.is_some());
    }

    #[test]
    fn test_compile_simple() {
        let source = "fn main() { let x = 42; }";
        let result = Compiler::compile(source);
        assert!(result.success || result.token_count > 0);
    }

    #[test]
    fn test_compilation_result() {
        let result = CompilationResult {
            success: true,
            c_code: Some("int main() { return 0; }".to_string()),
            errors: vec![],
            warnings: vec![],
            token_count: 10,
            ast_node_count: 5,
        };

        assert!(result.success);
        assert_eq!(result.errors.len(), 0);
    }
}
